### PR TITLE
Add type SponsorshipEvent

### PR DESCRIPTION
**Pull request by CodeDiego15**
### Description

This pull request adds support for the `SponsorshipEvent` to the **go-github** library, addressing the issue described in [#2982](https://github.com/google/go-github/issues/2982).

#### Details

According to GitHub documentation ([SponsorshipEvent](https://docs.github.com/en/rest/overview/github-event-types?apiVersion=2022-11-28#sponsorshipevent)), the `SponsorshipEvent` is triggered when sponsorship changes occur, such as creation, cancellation, or tier changes.

This PR introduces the following changes:
- **New `SponsorshipEvent` struct**: Captures the key properties of a sponsorship event, including:
  - `Action`: The type of action performed (e.g., `created`, `tier_changed`, `cancelled`, etc.).
  - `EffectiveDate`: The date when the event or change becomes effective, applicable for events like `pending_cancellation` and `pending_tier_change`.
  - `Changes`: An optional field that captures details about changes, such as adjustments in sponsorship tier or privacy. This is represented by an embedded struct called `SponsorshipChanges`.
  - Standard fields like `Repository`, `Organization`, `Sender`, and `Installation` are also included to maintain consistency with other GitHub event types.
  
- **New `SponsorshipChanges` struct**: This struct provides additional details when the `tier` or `privacy_level` of a sponsorship is modified.
  - The `Tier` field captures tier changes using an embedded `SponsorshipTier` struct, which contains a `From` field representing the previous tier.
  - The `PrivacyLevel` field tracks any changes to the sponsorship’s privacy settings.

This new event type will allow **go-github** users to programmatically handle sponsorship-related events triggered on GitHub, improving support for GitHub Sponsors workflows.

#### Testing
- Unit tests have been added to verify the correct deserialization of the `SponsorshipEvent` struct from JSON payloads.
- The tests cover various event scenarios, including tier changes, privacy level changes, and sponsorship creation.

#### Issue Link
Resolves [#2982](https://github.com/google/go-github/issues/2982)

### CODE REVIEW

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @CodeDiego15 !
Please read step 4 of CONTRIBUTING.md and push (not force-push) the resulting changes to this PR.

**Comment by CodeDiego15** [comment]
Force the push? Oh sorry I didn't notice.

**Comment by gmlewis** [comment]
> Force the push? Oh sorry I didn't notice.

No. Please push the results.  (Not a big deal, but please do _NOT_ use force push in this repo, as explained in CONTRIBUTING.md, as we always use squash and merge in this repo.)

**Comment by CodeDiego15** [comment]
I did a normal git push, I don't know what happened, I'm sorry

**Comment by gmlewis** [comment]
> Please read step 4 of CONTRIBUTING.md

Once again, please read step 4 of CONTRIBUTING.md.

Specifically,

> 4. Run `script/fmt.sh`, `script/test.sh` and `script/lint.sh` to format your code and
   check that it passes all tests and linters. `script/lint.sh` may also tell you
   that generated files need to be updated. If so, run `script/generate.sh` to
   update them.

**Comment by CodeDiego15** [comment]
That's what I'm trying right now, but when I do I get this: ./script/fmt.sh                                                                                                                
usage: dirname path

**Comment by gmlewis** [comment]
OK, then just try `./script/generate.sh`

**Comment by gmlewis** [comment]
And if that fails, then please try `go generate ./...`

**Comment by CodeDiego15** [comment]
That's it, some dependencies were downloaded, After that and excuse me. First time I contribute to the

**Comment by CodeDiego15** [comment]
Excuse me for bothering you so much, is there something missing to solve it?

**Comment by gmlewis** [comment]
Once you run `go generate ./...` then some auto-generated files should be updated.

Then, you can commit the changes, push to this PR, and the tests should all pass.

**Comment by gmlewis** [comment]
For example:

```bash
go generate ./...
git commit -sam 'run step 4'
git push origin add-sponsorship-event-type
```

**Comment by CodeDiego15** [comment]
I already did👍🏻

**Comment by CodeDiego15** [comment]
Was it done again as Git Push --forze? Because I did a normal git push?

**Review by gmlewis — CHANGES_REQUESTED** [review]
Please update `messages.go` by adding this somewhere around line 104:

```go
		"security_and_analysis":          &SecurityAndAnalysisEvent{},
		"sponsorship":  &SponsorshipEvent{},
		"star":                           &StarEvent{},
```

and also update `messages_test.go` respectively.  Thanks.

**Comment by CodeDiego15** [comment]
Hello good evening, an apology I had to leave. I'm already working on it

File: github/messages_test.go [hunk 1]
```diff
@@ -476,6 +476,10 @@ func TestParseWebHook(t *testing.T) {
 			payload:     &StatusEvent{},
 			messageType: "status",
 		},
+		{
```
Comment (gmlewis) [inline github/messages_test.go:479]: Please keep these sorted by moving these four lines up between "Security..." and "Star..." just like we did in `messages.go`.  Thanks.

Comment (CodeDiego15) [reply github/messages_test.go:479]: Like this? `{
			payload:     &SecurityAndAnalysisEvent{},
			messageType: "security_and_analysis",
		},
		{
			payload:     &StatusEvent{},
			messageType: "status",
		},
		{
			payload:     &StarEvent{},
			messageType: "star",
		},` and excuse mistakes

Comment (gmlewis) [reply github/messages_test.go:479]: No.

Like this:

```go
		{
			payload:     &SecretScanningAlertEvent{},
			messageType: "secret_scanning_alert",
		},
		{
			payload:     &SecurityAdvisoryEvent{},
			messageType: "security_advisory",
		},
		{
			payload:     &SecurityAndAnalysisEvent{},
			messageType: "security_and_analysis",
		},
		{
			payload:     &SponsorshipEvent{},
			messageType: "sponsors",
		},
		{
			payload:     &StarEvent{},
			messageType: "star",
		},
```

**Comment by gmlewis** [comment]
Still not sorted.  Please try again.

**Comment by CodeDiego15** [comment]
AAAA

**Comment by CodeDiego15** [comment]
I'm sorry, I don't know what to do anymore.

**Comment by gmlewis** [comment]
No problem.  I'll take over from here.  Thanks.

**Comment by gmlewis** [comment]
@CodeDiego15 -

Here's what you had before I made changes:

before-2024-09-09_20-41-58

Here's what I changed it to:

after-2024-09-09_20-45-11

See the difference?

It _always_ helps to take a look at the changed directly in the GitHub user interface so that you can see _exactly_ what is being changed.

**Review by gmlewis — APPROVED** [review]
Thank you, @CodeDiego15 !
LGTM.
Merging.

**Comment by CodeDiego15** [comment]
Thank you very much for accepting the PR will probably contribute to more issues.

**Pull request merged**